go/cmd/vtctldclient/internal/command/vschemas.go

```diff
@@ -33,8 +39,117 @@ var (
 		Args: cobra.ExactArgs(1),
 		RunE: commandGetVSchema,
 	}
+	// ApplyVSchema makes an ApplyVSchema gRPC call to a vtctld.
+	ApplyVSchema = &cobra.Command{
+		Use:  "ApplyVSchema {-vschema=<vschema> || -vschema_file=<vschema file> || -sql=<sql> || -sql_file=<sql file>} [-cells=c1,c2,...] [-skip_rebuild] [-dry-run] <keyspace>",
+		Args: cobra.ExactArgs(1),
+		RunE: commandApplyVSchema,
+	}
 )
 
+var applyVSchemaOptions = struct {
+	VSchema     string
+	VSchemaFile string
+	SQL         string
+	SQLFile     string
+	DryRun      bool
+	SkipRebuild bool
+	Cells       []string
+}{}
+
+func commandApplyVSchema(cmd *cobra.Command, args []string) error {
+	keyspace := cmd.Flags().Arg(0) // validated on the server-side
+
+	var vs *vschemapb.Keyspace
+	var err error
+
+	sqlMode := (applyVSchemaOptions.SQL != "") != (applyVSchemaOptions.SQLFile != "")
+	jsonMode := (applyVSchemaOptions.VSchema != "") != (applyVSchemaOptions.VSchemaFile != "")
+
+	if sqlMode && jsonMode {
+		return fmt.Errorf("only one of the sql, sql_file, vschema, or vschema_file flags may be specified when calling the ApplyVSchema command")
+	}
+
+	if !sqlMode && !jsonMode {
+		return fmt.Errorf("one of the sql, sql_file, vschema, or vschema_file flags must be specified when calling the ApplyVSchema command")
+	}
+
+	if sqlMode {
+		if applyVSchemaOptions.SQLFile != "" {
+			sqlBytes, err := ioutil.ReadFile(applyVSchemaOptions.SQLFile)
+			if err != nil {
+				return err
+			}
+			applyVSchemaOptions.SQL = string(sqlBytes)
+		}
+
+		stmt, err := sqlparser.Parse(applyVSchemaOptions.SQL)
+		if err != nil {
+			return fmt.Errorf("error parsing VSchema statement `%s`: %v", applyVSchemaOptions.SQL, err)
+		}
+		ddl, ok := stmt.(*sqlparser.AlterVschema)
+		if !ok {
+			return fmt.Errorf("error parsing VSchema statement `%s`: not a ddl statement", applyVSchemaOptions.SQL)
+		}
+
+		resp, err := client.GetVSchema(commandCtx, &vtctldatapb.GetVSchemaRequest{
+			Keyspace: keyspace,
+		})
+		if err != nil && !topo.IsErrType(err, topo.NoNode) {
+			return err
+		} // otherwise, we keep the empty vschema object from above
+
+		vs, err = topotools.ApplyVSchemaDDL(keyspace, resp.VSchema, ddl)
+		if err != nil {
+			return err
+		}
+
+	} else { // jsonMode
+		var schema []byte
+		if applyVSchemaOptions.VSchemaFile != "" {
+			schema, err = ioutil.ReadFile(applyVSchemaOptions.VSchemaFile)
+			if err != nil {
+				return err
+			}
+		} else {
+			schema = []byte(applyVSchemaOptions.VSchema)
+		}
+
+		vs = &vschemapb.Keyspace{}
+		err = json2.Unmarshal(schema, vs)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
interesting, i can definitely imagine someone calling the new api with only an sql string rather than a vschema object. so i can see where you're coming from.
i think i liked the idea of keeping the flag validation on the client side up above the call to FinishedParsing.
(though generally i do agree with what you say below about doing as much validation on the server side as possible.)

i do need to keep the file reading parts on the client side though, obviously. so it's not going to be a complete cut+paste job. but i'm definitely happy to shift the bulk of the logic up to the server side.